components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java

```diff
@@ -1100,6 +1107,112 @@ private void handleRoleClaim(AuthenticationContext context, Map<String, String>
+
+        // IdP claim mappings should be available and the current step should be a subject attribute step.
+        if (idPClaimMappings == null || !stepConfig.isSubjectAttributeStep()) {
+            return;
+        }
+        // Get the remote claim URI of the application role claim.
+        String idpAppRoleClaimUri = Arrays.stream(idPClaimMappings)
+                .filter(claimMapping -> claimMapping.getLocalClaim().getClaimUri()
+                        .equals(FrameworkConstants.APP_ROLES_CLAIM))
+                .map(claimMapping -> claimMapping.getRemoteClaim().getClaimUri())
+                .findFirst()
+                .orElse(null);
+        // If there is no application role claim mapping, no need to proceed.
+        if (StringUtils.isBlank(idpAppRoleClaimUri)) {
+            return;
+        }
+        // Regardless of whether the application role claim is requested from the SP, we need to add it to the remote
+        // claims since otherwise we wouldn't know if application roles are resolved or not at a later stage.
+        AppRoleMappingConfig[] appRoleMappingConfigs = context.getSequenceConfig().getApplicationConfig()
+                .getServiceProvider().getApplicationRoleMappingConfig();
+        String fidpName = stepConfig.getAuthenticatedIdP();
+        boolean useAppRoleMapping = Arrays.stream(appRoleMappingConfigs)
+                .anyMatch(appRoleMappingConfig -> appRoleMappingConfig.getIdPName().equals(fidpName)
+                        && appRoleMappingConfig.isUseAppRoleMappings());
+        if (useAppRoleMapping) {
+            String appRoles = getApplicationRoles(stepConfig.getAuthenticatedUser(), context);
+            // Checking if the appRoles string is null but can be an empty string.
+            if (appRoles != null) {
+                remoteClaims.put(idpAppRoleClaimUri, appRoles);
+            }
+        }
+    }
+
+    /**
+     * Add the application roles of local user to local claims.
+     *
+     * @param stepConfig StepConfig of current step.
+     * @param context AuthenticationContext of current authentication flow.
+     * @param allLocalClaims All local claims of the current authenticated user.
+     * @throws FrameworkException Exception on handling application roles for local user.
+     */
+    protected void handleApplicationRolesForLocalUser(StepConfig stepConfig, AuthenticationContext context,
+                                                    Map<String, String> allLocalClaims)
+            throws FrameworkException {
+
+        AuthenticatedUser authenticatedUser = (stepConfig == null)
+                ? context.getSequenceConfig().getAuthenticatedUser()
+                : stepConfig.getAuthenticatedUser();
+        if (stepConfig == null || stepConfig.isSubjectAttributeStep()) {
+            String requestedAppRoleClaim = context.getSequenceConfig().getApplicationConfig()
+                    .getRequestedClaimMappings().get(FrameworkConstants.APP_ROLES_CLAIM);
```

Comment:
are we sure that non of the intermediate returned value of `context.getSequenceConfig().getApplicationConfig()
                    .getRequestedClaimMappings()` are not null?
If null value if possible check not null condition prior invoking a method on the object

Reply:
In this instance, there would not be any null values in returned values as checked.
However added null checks for other places with 734a7bff58df0ecbf63400a8d7f3f61cf3ff22b4